modules/fedimint-ecash-migration-server/src/lib.rs

```diff
@@ -227,20 +435,321 @@ impl ServerModule for EcashMigration {
+                ApiVersion::new(0, 0),
+                async |module: &EcashMigration, context, request: TransferId| -> u64 {
+                    module.get_uploaded_spend_book_entries(&mut context.dbtx().into_nc(), request)
+                        .await
+                }
+            },
+            api_endpoint! {
+                CHECK_SPEND_BOOK_HASH_ENDPOINT,
+                ApiVersion::new(0, 0),
+                async |module: &EcashMigration, context, request: TransferId| -> SpendBookHash {
+                    // TODO: check auth
+                    module.check_spend_book_hash(&mut context.dbtx().into_nc(), request)
+                        .await
+                }
+            },
+            api_endpoint! {
+                GET_TRANSFER_STATUS_ENDPOINT,
+                ApiVersion::new(0, 0),
+                async |module: &EcashMigration, context, request: GetTransferStatusRequest| -> GetTransferStatusResponse {
+                    module.get_transfer_status(&mut context.dbtx().into_nc(), request.transfer_id)
+                        .await
+                }
+            },
+        ]
     }
 }
 
 impl EcashMigration {
-    /// Create new module instance
-    pub fn new(cfg: EcashMigrationConfig) -> EcashMigration {
-        EcashMigration { cfg }
+    /// Returns the transfer ID of the transfer created by the output
+    async fn get_transfer_id(
+        &self,
+        dbtx: &mut DatabaseTransaction<'_>,
+        out_point: OutPoint,
+    ) -> Option<TransferId> {
+        dbtx.get_value(&OutPointTransferIdKey(out_point)).await
+    }
+
+    async fn upload_key_set(
+        &self,
+        dbtx: &mut DatabaseTransaction<'_>,
+        transfer_id: TransferId,
+        key_set: Tiered<AggregatePublicKey>,
+    ) -> Result<(), ApiError> {
+        let transfer_metadata = dbtx
+            .get_value(&TransferMetadataKey(transfer_id))
+            .await
+            .ok_or(ApiError::not_found("Transfer not found".to_owned()))?;
+
+        let key_set_hash = key_set.consensus_hash_sha256();
+        if key_set_hash != transfer_metadata.origin_key_set_hash.0 {
+            return Err(ApiError::bad_request(
+                "Key set hash does not match pre-committed hash".to_owned(),
+            ));
+        }
+
+        for (amount, public_key) in key_set {
+            if dbtx
+                .insert_entry(
+                    &DenominationKeyKey {
+                        transfer_id,
+                        amount,
+                    },
+                    &public_key,
+                )
+                .await
+                .is_some()
+            {
+                // We can assume that finding a collision of the key set hash is impossible,
+                // hence if we see one of the keys we can assume all the other keys are also
+                // present.
+                return Ok(());
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn upload_spend_book_batch(
```

Comment:
So in no way is this authenticated and until we get all entries we can't tell if there was no "corruption"? :/